### ISSUE TITLE
Fatal: No uct_ep_h associated with ucp_ep_h on ucp_worker_h 

### ISSUE THREAD

**Issue by alinask** [openucx/ucx#1578]
The following failure reproduces on every run and can be triggered with the following command line.

It runs with 2 HCAs: UCX_NET_DEVICES=mlx5_2:1,mlx5_0:1 but passes on each device separately.
On the cluster, mlx5_2 is ConnectX-5 and mlx5_0 is ConnectX-4.
These are 24 hosts. Passes with 256 ranks but doesn’t with 512 and above.
```
/hpc/local/benchmarks/hpcx_install_Monday/hpcx-gcc-redhat7.2/ompi-v2.x/bin/mpirun -np 768 \
   -mca btl_openib_warn_default_gid_prefix 0  --bind-to core --tag-output --timestamp-output  \
   -mca pml ucx -x UCX_NET_DEVICES=mlx5_2:1,mlx5_0:1 -mca btl_openib_if_include mlx5_0:1\
   -mca coll_hcoll_enable 0 -x UCX_TLS=rc_x,sm -mca opal_pmix_base_async_modex 1 \
   -mca mpi_add_procs_cutoff 0 -mca pmix_base_collect_data 0 --map-by node \
   /hpc/scrap/users/mtt/scratch/ucx_ompi/20170605_220537_22627_746227_clx-hercules-012/installs/rxQ4/tests/mpich_tests/mpich-mellanox.git/test/mpi/datatype/unpack 
```
http://e2e-gw.mellanox.com:4080/hpc/scrap/users/mtt/scratch/ucx_ompi/20170605_220537_22627_746227_clx-hercules-012/html/test_stdout_iOkI3t.txt

A trace from the failure is attached.

**Comment by alinask** [comment]
[1578_trace.txt](https://github.com/openucx/ucx/files/1055737/1578_trace.txt)

**Comment by alinask** [comment]
One more command line to reproduce (a different benchmark)

/hpc/local/benchmarks/hpcx_install_Tuesday/hpcx-gcc-redhat7.2/ompi-v2.x/bin/mpirun -np 1280 -mca btl_openib_warn_default_gid_prefix 0 --bind-to core --tag-output --timestamp-output --display-map -mca pml ucx -x UCX_NET_DEVICES=mlx5_0:1,mlx5_1:1,mlx5_2:1,mlx5_3:1 -mca btl_openib_if_include mlx5_0:1,mlx5_1:1,mlx5_2:1,mlx5_3:1 -mca coll_hcoll_enable 0 -x UCX_TLS=rc_x,sm -mca opal_pmix_base_async_modex 0 -mca mpi_add_procs_cutoff 100000 --map-by node /hpc/scrap/users/mtt/scratch/ucx_ompi/20170606_184804_3438_746638_clx-hercules-073/installs/DzlD/tests/mpich_tests/mpich-mellanox.git/test/mpi/datatype/simple-resized

```
Tue Jun  6 19:14:48 2017[1,639]<stderr>:[clx-hercules-112:20143:0]  ucp_worker.c:260  Fatal: No uct_ep_h 0xafc140 associated with ucp_ep_h on ucp_worker_h 0x7ffff7ea0010
Tue Jun  6 19:14:48 2017[1,617]<stderr>:[clx-hercules-090:32652:0]  ucp_worker.c:260  Fatal: No uct_ep_h 0xbe8400 associated with ucp_ep_h on ucp_worker_h 0x7ffff7e96010
Tue Jun  6 19:14:48 2017[1,619]<stderr>:[clx-hercules-092:5674 :0]  ucp_worker.c:260  Fatal: No uct_ep_h 0xb751f0 associated with ucp_ep_h on ucp_worker_h 0x7ffff7e96010
Tue Jun  6 19:14:48 2017[1,588]<stderr>:[clx-hercules-101:10199:0]  ucp_worker.c:260  Fatal: No uct_ep_h 0xb7d240 associated with ucp_ep_h on ucp_worker_h 0x7ffff7ea0010
Tue Jun  6 19:14:48 2017[1,1243]<stderr>:[clx-hercules-076:28600:0]  ucp_worker.c:260  Fatal: No uct_ep_h 0xbce700 associated with ucp_ep_h on ucp_worker_h 0x7ffff7e8c010
Tue Jun  6 19:14:48 2017[1,639]<stderr>:==== backtrace ====
Tue Jun  6 19:14:48 2017[1,639]<stderr>: 0 0x000000000000ed32 ucp_worker_iface_error_handler()  /hpc/local/benchmarks/hpcx_install_Tuesday/src/hpcx-gcc-redhat7.2/ucx-master/src/ucp/core/ucp_worker.c:259
Tue Jun  6 19:14:48 2017[1,639]<stderr>: 1 0x00000000000379f0 uct_ud_iface_dispatch_async_comps_do()  /hpc/local/benchmarks/hpcx_install_Tuesday/src/hpcx-gcc-redhat7.2/ucx-master/src/uct/ib/ud/base/ud_iface.c:696
Tue Jun  6 19:14:48 2017[1,639]<stderr>: 2 0x000000000003e2d6 uct_ud_iface_dispatch_zcopy_comps()  /hpc/local/benchmarks/hpcx_install_Tuesday/src/hpcx-gcc-redhat7.2/ucx-master/src/uct/ib/ud/base/ud_iface.h:390
Tue Jun  6 19:14:48 2017[1,639]<stderr>: 3 0x000000000003e2d6 uct_ud_mlx5_iface_progress()  /hpc/local/benchmarks/hpcx_install_Tuesday/src/hpcx-gcc-redhat7.2/ucx-master/src/uct/ib/ud/accel/ud_mlx5.c:445
Tue Jun  6 19:14:48 2017[1,639]<stderr>: 4 0x000000000001989e ucs_callbackq_dispatch()  /hpc/local/benchmarks/hpcx_install_Tuesday/src/hpcx-gcc-redhat7.2/ucx-master/src/ucs/datastruct/callbackq.inl:39
Tue Jun  6 19:14:48 2017[1,639]<stderr>: 5 0x000000000001989e uct_worker_progress()  /hpc/local/benchmarks/hpcx_install_Tuesday/src/hpcx-gcc-redhat7.2/ucx-master/src/uct/base/uct_md.c:233
Tue Jun  6 19:14:48 2017[1,639]<stderr>: 6 0x000000000000fbad ucp_worker_progress()  /hpc/local/benchmarks/hpcx_install_Tuesday/src/hpcx-gcc-redhat7.2/ucx-master/src/ucp/core/ucp_worker.c:812
Tue Jun  6 19:14:48 2017[1,639]<stderr>: 7 0x000000000000fbad ucs_async_check_miss()  /hpc/local/benchmarks/hpcx_install_Tuesday/src/hpcx-gcc-redhat7.2/ucx-master/src/ucs/async/async.h:75
Tue Jun  6 19:14:48 2017[1,639]<stderr>: 8 0x000000000000fbad ucp_worker_progress()  /hpc/local/benchmarks/hpcx_install_Tuesday/src/hpcx-gcc-redhat7.2/ucx-master/src/ucp/core/ucp_worker.c:813
Tue Jun  6 19:14:48 2017[1,639]<stderr>: 9 0x0000000000002f97 mca_pml_ucx_progress()  /hpc/local/benchmarks/hpcx_install_Tuesday/src/hpcx-gcc-redhat7.2/ompi-v2.x/ompi/mca/pml/ucx/pml_ucx.c:421
Tue Jun  6 19:14:48 2017[1,639]<stderr>:10 0x000000000002ef2c opal_progress()  /hpc/local/benchmarks/hpcx_install_Tuesday/src/hpcx-gcc-redhat7.2/ompi-v2.x/opal/runtime/opal_progress.c:225
Tue Jun  6 19:14:48 2017[1,639]<stderr>:11 0x000000000000403d mca_pml_ucx_waitall()  /hpc/local/benchmarks/hpcx_install_Tuesday/src/hpcx-gcc-redhat7.2/ompi-v2.x/ompi/mca/pml/ucx/pml_ucx.c:340
Tue Jun  6 19:14:48 2017[1,639]<stderr>:12 0x000000000000403d mca_pml_ucx_del_procs()  /hpc/local/benchmarks/hpcx_install_Tuesday/src/hpcx-gcc-redhat7.2/ompi-v2.x/ompi/mca/pml/ucx/pml_ucx.c:396
Tue Jun  6 19:14:48 2017[1,639]<stderr>:13 0x000000000004511d ompi_mpi_finalize()  /hpc/local/benchmarks/hpcx_install_Tuesday/src/hpcx-gcc-redhat7.2/ompi-v2.x/ompi/runtime/ompi_mpi_finalize.c:318
Tue Jun  6 19:14:48 2017[1,639]<stderr>:14 0x00000000004026b6 main()  ???:0
Tue Jun  6 19:14:48 2017[1,639]<stderr>:15 0x0000000000021b15 __libc_start_main()  ???:0
Tue Jun  6 19:14:48 2017[1,639]<stderr>:16 0x0000000000402539 _start()  ???:0
Tue Jun  6 19:14:48 2017[1,639]<stderr>:===================
```

http://e2e-gw.mellanox.com:4080/hpc/scrap/users/mtt/scratch/ucx_ompi/20170606_184804_3438_746638_clx-hercules-073/html/test_stdout_3l4Eq1.txt

**Comment by yosefe** [comment]
Fixed by #1594
